### PR TITLE
Dance: add soft button support, fix key event handling

### PR DESCRIPTION
**Pull request by cpirich**
* Migrated GameLab's support for soft buttons over to Dance: `ButtonState`, `ArrowIds`, the event handlers, the icons themselves and the CSS that enables them. There is some duplication with GameLab (and, to a lesser extent, with other apps), but there are also some differences in Dance (no d-pad, not separate object that wraps `this.p5`) that make it non-trivial to try to share this code.
* Enabled soft buttons on the Dance project level

### CODE REVIEW

File: apps/src/dance/Dance.js [hunk 1]
```diff
@@ -109,6 +125,7 @@ Dance.prototype.init = function (config) {
   };
 
   const showFinishButton = !this.level.isProjectLevel && !this.level.validationCode;
+  const finishButtonFirstLine = _.isEmpty(this.level.softButtons);
```
Comment (joshlory) [inline apps/src/dance/Dance.js:128]:
```suggestion
  const finishButtonFirstLine = Object.keys(this.level.softButtons).length === 0;
```

File: apps/src/dance/Dance.js [hunk 2]
```diff
@@ -141,10 +160,83 @@ Dance.prototype.loadAudio_ = function () {
...
+    if (this.btnState[buttonId] === ButtonState.DOWN) {
+      this.onArrowButtonUp(buttonId, e);
+    }
+  }
+};
+
+Dance.prototype.notifyKeyCodeDown = function (keyCode) {
+  // Synthesize an event and send it to the internal p5 handler for keydown
+  if (this.p5) {
+    this.p5._onkeydown({ which: keyCode });
```
Comment (joshlory) [inline apps/src/dance/Dance.js:209]: Is there a way to dispatch this that doesn't require reaching in to private APIs?

File: apps/src/dance/Dance.js [hunk 3]
```diff
@@ -176,6 +268,15 @@ Dance.prototype.reset = function () {
...
   this.nativeAPI.reset();
   this.p5.noLoop();
+
+  var softButtonCount = 0;
+  for (var i = 0; i < this.level.softButtons.length; i++) {
+    document.getElementById(this.level.softButtons[i]).style.display = 'inline';
+    softButtonCount++;
+  }
+  if (softButtonCount) {
+    $('#soft-buttons').removeClass('soft-buttons-none').addClass('soft-buttons-' + softButtonCount);
```
Comment (joshlory) [inline apps/src/dance/Dance.js:278]: Nit: this has an implicit dependency on `window.$` (vs. an explicit `import` statement).  Worth writing this without jQuery?

File: apps/style/dance/style.scss [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  margin-right: 3px;
+}
+.up-btn {
+  background-position: 0px -63px;
+  margin-right: 3px;
+}
+.down-btn {
+  background-position: 0px -42px;
+  margin-right: 3px;
+}
```
Comment (joshlory) [inline apps/style/dance/style.scss:53]: Seeing this duplicated all around the codebase, worth extracting out a shared buttons.scss include?

File: apps/src/dance/Dance.js (hunk 2 above)
Comment (cpirich) [reply apps/src/dance/Dance.js:209]: p5 was really intended to allow this. I'm not aware of another alternative approach. Synthesizing the key events with the expectation that p5 will see them and handle them identically to real key events seems riskier with various browsers.

**Comment by cpirich** [comment]
Some good feedback. For now, I'm going to stick with the exact code we've been using in Game Lab just to get things working again. Refactoring common code and cleaning up some of the older patterns (e.g. lowdash/underscore) will need to wait. I did make a few changes to use `const` and remove the pieces that are not applicable to Dance, but tried to keep the runtime code that we're executing identical to the existing, tested code paths from Game Lab.

**Pull request merged**